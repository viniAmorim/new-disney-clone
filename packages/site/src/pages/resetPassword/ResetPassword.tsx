import React, { BaseHTMLAttributes } from 'react';
import { UseFormReturn } from 'react-hook-form';
import { Col, Container, Row } from 'react-grid-system';
import { FaPlus } from 'react-icons/fa';

import { FormControl, Input, Heading, Button, Paragraph } from '@site/uikit';
import Link from 'next/link';

import * as Styled from './styles';
import { ResetPasswordFormValues } from './ResetPasswordContainer';

interface ResetPasswordProps extends BaseHTMLAttributes<HTMLDivElement> {
  onResetPassword: (values: ResetPasswordFormValues) => void;
  form: UseFormReturn;
  isLoading?: boolean;
}

function ResetPassword({
  onResetPassword,
  form,
  isLoading,
  ...props
}: ResetPasswordProps) {
  return (
    <Styled.Background {...props}>
      <Container>
        <Row>
          <Col sm={6} md={6}>
            <Styled.Logo src="/images/logo.png" alt="Correta.app" />
            <Styled.StyledHeading size="headline">
              Nova senha
            </Styled.StyledHeading>
            <form onSubmit={form.handleSubmit(onResetPassword)}>
              <FormControl
                label="E-mail:"
                error={form.formState.errors?.email?.message}
              >
                <Input
                  name="email"
                  placeholder="seu@email.com"
                  {...form.register('email', {
                    required: true,
                    pattern: {
                      value: /\S+@\S+\.\S+/,
                      message: 'E-mail inválido.',
                    },
                    minLength: {
                      value: 5,
                      message: 'E-mail é obrigatório',
                    },
                  })}
                />
              </FormControl>
              <FormControl
                label="Código:"
                error={form.formState.errors?.code?.message}
              >
                <Input
                  name="code"
                  placeholder="Código recebido no e-mail"
                  {...form.register('code', {
                    required: true,
                    minLength: {
                      value: 2,
                      message: 'Código é obrigatório',
                    },
                  })}
                />
              </FormControl>
              <FormControl
                label="Nova Senha:"
                error={form.formState.errors?.password?.message}
              >
                <Input
                  name="password"
                  type="password"
                  placeholder="password"
                  {...form.register('password', {
                    required: true,
                    minLength: {
                      value: 5,
                      message: 'Senha é obrigatório',
                    },
                  })}
                  data-testid="password-input"
                />
              </FormControl>
              <FormControl
                label="Repetir Senha:"
                error={form.formState.errors?.repeatPassword?.message}
              >
                <Input
                  name="repeatPassword"
                  type="password"
                  placeholder="password"
                  {...form.register('repeatPassword', {
                    required: true,
                    minLength: {
                      value: 5,
                      message: 'Repetir Senha é obrigatório',
                    },
                    validate: value => {
                      if (value === form.getValues()['password']) {
                        return true;
                      } else {
                        return 'Senha e Repetir senha precisam ser iguais.';
                      }
                    },
                  })}
                  data-testid="repeat-password-input"
                />
              </FormControl>
              <Styled.ResetPasswordContainer>
                <Button
                  data-testid="reset-button"
                  type="submit"
                  isLoading={isLoading}
                >
                  Criar nova senha
                </Button>
              </Styled.ResetPasswordContainer>
              <Styled.RegisterContainer>
                <Paragraph size="small">Ainda não tem cadastro?</Paragraph>
                <Link href="/register">
                  <Button variant="ghost">Cadastre-se gratuitamente</Button>
                </Link>
              </Styled.RegisterContainer>
            </form>
          </Col>
          <Styled.ContentCol sm={6} md={6}>
            <Styled.ContentContainer>
              <img src="/images/home.png" alt="Correta.app" />
              <span>
                <FaPlus />
                <Heading size="display">facilidade</Heading>
              </span>
              <span>
                <FaPlus />
                <Heading size="display">resultados</Heading>
              </span>
              <Paragraph size="headline">
                Tenha acesso e controle de seus cliente, melhorando seu processo
                de vendas e gerando mais resultados.
              </Paragraph>
            </Styled.ContentContainer>
          </Styled.ContentCol>
        </Row>
      </Container>
    </Styled.Background>
  );
}

export default ResetPassword;
