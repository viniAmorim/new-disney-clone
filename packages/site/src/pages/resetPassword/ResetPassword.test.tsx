import React from 'react';

import userEvent from '@testing-library/user-event';

import { fireEvent, render, waitFor } from '~/test/utils';

import ResetPassword from './ResetPassword';

const onResetPassword = jest.fn();

describe('ResetPassword', () => {
  it('should render correctly', async () => {
    const { queryByTestId } = render(
      <ResetPassword data-testid="home" onResetPassword={onResetPassword} />,
    );

    const home = queryByTestId('home');

    expect(home).toBeInTheDocument();
    expect(home).toMatchSnapshot();
  });

  it('should show error when email is empty', async () => {
    const { getByTestId, getByText } = render(
      <ResetPassword onResetPassword={onResetPassword} />,
    );

    const button = getByTestId('reset-button');
    fireEvent.click(button);

    await waitFor(() => expect(getByText('E-mail é obrigatório')).toBeTruthy());
  });

  it('should show error when email is invalid', async () => {
    const { getAllByRole, getByTestId, getByText } = render(
      <ResetPassword onResetPassword={onResetPassword} />,
    );

    const [emailInput] = getAllByRole('textbox');
    userEvent.type(emailInput, 'invalid-email.com');

    const button = getByTestId('reset-button');
    fireEvent.click(button);

    await waitFor(() => expect(getByText('E-mail inválido')).toBeTruthy());
  });

  it('should show error when code is empty', async () => {
    const { getAllByRole, getByTestId, getByText } = render(
      <ResetPassword onResetPassword={onResetPassword} />,
    );

    const [emailInput] = getAllByRole('textbox');
    userEvent.type(emailInput, 'email@email.com');

    const button = getByTestId('reset-button');
    fireEvent.click(button);

    await waitFor(() => expect(getByText('Código é obrigatório')).toBeTruthy());
  });

  it('should show error when password is empty', async () => {
    const { getAllByRole, getByTestId, getByText } = render(
      <ResetPassword onResetPassword={onResetPassword} />,
    );

    const [emailInput, codeInput] = getAllByRole('textbox');
    userEvent.type(emailInput, 'email@email.com');
    userEvent.type(codeInput, 'code');

    const button = getByTestId('reset-button');
    fireEvent.click(button);

    await waitFor(() => expect(getByText('Senha é obrigatória')).toBeTruthy());
  });

  it('should show error when repeat password is empty', async () => {
    const { getAllByRole, getByTestId, getByText } = render(
      <ResetPassword onResetPassword={onResetPassword} />,
    );

    const [emailInput, codeInput] = getAllByRole('textbox');
    userEvent.type(emailInput, 'email@email.com');
    userEvent.type(codeInput, 'code');

    const passwordInput = getByTestId('password-input');
    userEvent.type(passwordInput, 'password');

    const button = getByTestId('reset-button');
    fireEvent.click(button);

    await waitFor(() =>
      expect(getByText('Repetir Senha é obrigatória')).toBeTruthy(),
    );
  });

  it('should show error when repeat password is different from password field', async () => {
    const { getAllByRole, getByTestId, getByText } = render(
      <ResetPassword onResetPassword={onResetPassword} />,
    );

    const [emailInput, codeInput] = getAllByRole('textbox');
    userEvent.type(emailInput, 'email@email.com');
    userEvent.type(codeInput, 'code');

    const passwordInput = getByTestId('password-input');
    userEvent.type(passwordInput, 'password');

    const repeatPasswordInput = getByTestId('repeat-password-input');
    userEvent.type(repeatPasswordInput, 'password2');

    const button = getByTestId('reset-button');
    fireEvent.click(button);

    await waitFor(() =>
      expect(
        getByText('Senha e Repetir senha precisam ser iguais.'),
      ).toBeTruthy(),
    );
  });

  it('should get reset password data when clicked in reset button', async () => {
    const { getAllByRole, getByTestId } = render(
      <ResetPassword onResetPassword={onResetPassword} />,
    );

    const [emailInput, codeInput] = getAllByRole('textbox');
    userEvent.type(emailInput, 'email@email.com');
    userEvent.type(codeInput, 'code');

    const passwordInput = getByTestId('password-input');
    userEvent.type(passwordInput, 'password');

    const repeatPasswordInput = getByTestId('repeat-password-input');
    userEvent.type(repeatPasswordInput, 'password');

    const button = getByTestId('reset-button');
    fireEvent.click(button);

    await waitFor(() => {
      expect(onResetPassword.mock.calls[0][0]).toEqual({
        code: 'code',
        email: 'email@email.com',
        password: 'password',
        repeatPassword: 'password',
      });
    });
  });
});
