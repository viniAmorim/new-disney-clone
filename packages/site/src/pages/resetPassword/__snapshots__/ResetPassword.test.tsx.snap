// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPassword should render correctly 1`] = `
<div
  class="styles__Background-sc-1jm3niw-2 bppNnm"
  data-testid="home"
>
  <div
    style="box-sizing: border-box; position: relative; margin-left: auto; margin-right: auto; padding-left: 15px; padding-right: 15px; max-width: 960px;"
  >
    <div
      style="margin-left: -15px; margin-right: -15px; display: flex; flex-wrap: wrap; flex-grow: 0; flex-shrink: 0; align-items: normal; justify-content: flex-start; flex-direction: row;"
    >
      <div
        style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 50%; flex-basis: 50%; flex-grow: 0; flex-shrink: 0; max-width: 50%; margin-left: 0%;"
      >
        <img
          alt="Correta.app"
          class="styles__Logo-sc-1jm3niw-0 hKLbrS"
          src="/images/logo.png"
        />
        <h1
          class="sc-gKAaRy kbUCYm styles__StyledHeading-sc-1jm3niw-1 fPKxRr"
        >
          Nova senha
        </h1>
        <form>
          <div
            class="sc-hKFxyN zFDVE"
          >
            <label
              class="sc-eCApnc CDjog"
            >
              E-mail:
            </label>
            <input
              class="sc-iCoGMd kmeLFP"
              name="email"
              placeholder="seu@email.com"
              value=""
            />
          </div>
          <div
            class="sc-hKFxyN zFDVE"
          >
            <label
              class="sc-eCApnc CDjog"
            >
              Código:
            </label>
            <input
              class="sc-iCoGMd kmeLFP"
              name="code"
              placeholder="Código recebido no e-mail"
              value=""
            />
          </div>
          <div
            class="sc-hKFxyN zFDVE"
          >
            <label
              class="sc-eCApnc CDjog"
            >
              Nova Senha:
            </label>
            <input
              class="sc-iCoGMd kmeLFP"
              data-testid="password-input"
              name="password"
              placeholder="password"
              type="password"
              value=""
            />
          </div>
          <div
            class="sc-hKFxyN zFDVE"
          >
            <label
              class="sc-eCApnc CDjog"
            >
              Repetir Senha:
            </label>
            <input
              class="sc-iCoGMd kmeLFP"
              data-testid="repeat-password-input"
              name="repeatPassword"
              placeholder="password"
              type="password"
              value=""
            />
          </div>
          <div
            class="styles__ResetPasswordContainer-sc-1jm3niw-5 MgqXQ"
          >
            <button
              class="sc-gtsrHT kGzhap"
              data-testid="reset-button"
              type="submit"
            >
              Criar nova senha
            </button>
          </div>
          <div
            class="styles__RegisterContainer-sc-1jm3niw-6 fYsuHj"
          >
            <p
              class="sc-iemWCZ eTebHM"
            >
              Ainda não tem cadastro?
            </p>
            <button
              class="sc-gtsrHT gBmKmV"
            >
              Cadastre-se gratuitamente
            </button>
          </div>
        </form>
      </div>
      <div
        class="styles__ContentCol-sc-1jm3niw-3 FwtAp"
        style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 15px; padding-right: 15px; width: 50%; flex-basis: 50%; flex-grow: 0; flex-shrink: 0; max-width: 50%; margin-left: 0%;"
      >
        <div
          class="styles__ContentContainer-sc-1jm3niw-4 bJzkKl"
        >
          <img
            alt="Correta.app"
            src="/images/home.png"
          />
          <span>
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              />
            </svg>
            <h1
              class="sc-gKAaRy eeYKHx"
            >
              facilidade
            </h1>
          </span>
          <span>
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              />
            </svg>
            <h1
              class="sc-gKAaRy eeYKHx"
            >
              resultados
            </h1>
          </span>
          <p
            class="sc-iemWCZ bCVkDu"
          >
            Tenha acesso e controle de seus cliente, melhorando seu processo de vendas e gerando mais resultados.
          </p>
        </div>
      </div>
    </div>
    <span
      style="display: table; clear: both;"
    />
  </div>
</div>
`;
